File: libs/testing/src/fixtures.rs
//! Deterministic JWT issuance and SQL seed helpers for tests.
//!
//! The auth helpers reuse [`auth_middleware::jwt`] so the issued tokens
//! pass the production [`auth_middleware::layer::auth_layer`] without
//! any special-case handling.

use auth_middleware::jwt::{self, JwtConfig, build_access_claims, encode_token};
use serde_json::Value;
use sqlx::PgPool;
use uuid::Uuid;

/// Shared HS256 secret used across the test suite. Long enough to
/// satisfy the upstream length checks; never used outside tests.
pub const JWT_SECRET: &str = "openfoundry-shared-test-secret-do-not-use-in-prod-aaaa";

/// Build a [`JwtConfig`] from [`JWT_SECRET`].
pub fn jwt_config() -> JwtConfig {
    jwt::JwtConfig::new(JWT_SECRET)
}

/// Issue a Bearer-ready JWT with the requested permissions. The user id
/// is randomised so each call produces a distinct subject.
pub fn dev_token(cfg: &JwtConfig, permissions: Vec<String>) -> String {
    let claims = build_access_claims(
        cfg,
        Uuid::now_v7(),
        "tester@openfoundry.test",
        "Integration Tester",
        vec!["admin".to_string()],
        permissions,
        None,
        Value::Null,
        vec!["password".to_string()],
    );
    encode_token(cfg, &claims).expect("encode test token")
}

/// Insert a minimal `datasets` row, returning its id. Works against
/// either the catalog or versioning schemas (both use the same column
/// names: `id`, `rid`, `name`, `format`, `storage_path`, `owner_id`).
pub async fn seed_dataset(pool: &PgPool, rid: &str, name: &str, format: &str) -> Uuid {
    let id = Uuid::now_v7();
    sqlx::query(
        r#"INSERT INTO datasets (id, rid, name, format, storage_path, owner_id)
           VALUES ($1, $2, $3, $4, $5, $6)"#,
    )
    .bind(id)
    .bind(rid)
    .bind(name)
    .bind(format)
    .bind(format!("local://{rid}/v1"))
    .bind(Uuid::now_v7())
    .execute(pool)
    .await
    .expect("seed dataset row");
    id
}

File: openfoundry-go/libs/testing/fixtures.go
package testingx

import (
	"context"
	"encoding/json"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgxpool"

	authmw "github.com/openfoundry/openfoundry-go/libs/auth-middleware"
	"github.com/openfoundry/openfoundry-go/libs/core-models/ids"
)

// JWTSecret is the shared HS256 secret used across the integration
// test suite. Long enough for HS256 length checks; never used outside tests.
const JWTSecret = "openfoundry-shared-test-secret-do-not-use-in-prod-aaaa"

// JWTConfig builds an authmw.JWTConfig from JWTSecret.
func JWTConfig() *authmw.JWTConfig {
	return authmw.NewJWTConfig(JWTSecret)
}

// DevToken issues a Bearer-ready JWT with the requested permissions.
//
// The subject is randomised so each call produces a distinct user. The
// returned token passes the production Middleware without special-case
// handling — same contract as the Rust `dev_token` helper.
func DevToken(cfg *authmw.JWTConfig, permissions []string) (string, error) {
	now := time.Now()
	c := &authmw.Claims{
		Sub:         ids.New(),
		IAT:         now.Unix(),
		EXP:         now.Add(cfg.AccessTTL).Unix(),
		JTI:         ids.New(),
		Email:       "tester@openfoundry.test",
		Name:        "Integration Tester",
		Roles:       []string{"admin"},
		Permissions: permissions,
		Attributes:  json.RawMessage(`null`),
		AuthMethods: []string{"password"},
	}
	return authmw.EncodeToken(cfg, c)
}

// SeedDataset inserts a minimal `datasets` row and returns its id.
//
// Works against any schema that uses the canonical column set:
// (id, rid, name, format, storage_path, owner_id) — both the catalog
// and versioning schemas follow this convention.
func SeedDataset(ctx context.Context, pool *pgxpool.Pool, rid, name, format string) (uuid.UUID, error) {
	id := ids.New()
	_, err := pool.Exec(ctx,
		`INSERT INTO datasets (id, rid, name, format, storage_path, owner_id)
		 VALUES ($1, $2, $3, $4, $5, $6)`,
		id, rid, name, format, "local://"+rid+"/v1", ids.New(),
	)
	return id, err
}
